ext/hash/config.m4

```diff
@@ -33,7 +33,7 @@ else
   PHP_HASH_CFLAGS="$PHP_HASH_CFLAGS -I@ext_srcdir@/$SHA3_DIR -DKeccakP200_excluded -DKeccakP400_excluded -DKeccakP800_excluded -DZEND_ENABLE_STATIC_TSRMLS_CACHE=1"
 fi
 
-EXT_HASH_SOURCES="hash.c hash_md.c hash_sha.c hash_ripemd.c hash_haval.c \
+EXT_HASH_SOURCES="hash.c hash_md.c hash_sha.c hash_sha_sse2.c hash_sha_ni.c hash_ripemd.c hash_haval.c \
```

Comment:
It would be nice to have support for this on Windows, too; maybe just try to add these files to ext/hash/config.w32 and see what CI makes of it. At least SSE2 shouldn't be an issue on Windows, since the build system defaults to this for many years (CI even requires AVX2 instructions).

Reply:
I've made the suggested change and pushed. I don't use Windows myself, though. Please check yourself and push any necessary fixes into my branch (pushing is enabled for maintainers):

![image](https://github.com/user-attachments/assets/63fd870b-f367-42a1-86ad-d9ff2c651527)

Reply:
The problem are the `static restrict` qualifiers when some array arguments are declared. Removing these makes the build pass. I've pushed a quick fix, but according to https://stackoverflow.com/questions/53863084/what-optimization-benefit-does-pointerrestrict-static-1-bring-when-declare-a this construct is supported by C, but not by C++, so the `#ifdef` would need to be fixed. I'm not sure about that, though.

Reply:
Have you been able to verify if the execution on Windows has become faster / uses the SSE / SHA-NI implementation?

Reply:
A default build runs the SSE2 code branch. An AVX2 build still does for me (not sure if I have support for Intel SHA instructions). I'll check that, but first a couple of Windows build issues need to be resolved (e.g. `HAVE_IMMINTRISIC_H` is never defined, although that header file is apparently available (Windows SDK 10.0.20348.0).

Reply:
Ugh, that whole AVX detection is apparently pretty much messed up in php-src. Prior to this PR, `HAVE_IMMINTRISIC_H` is only used once in zend_portability.h, to conditionally define `PHP_HAVE_AVX2` (but not for MSVC), which is then used to conditionally define `ZEND_INTRIN_AVX2_RESOLVER` (which is always done on Windows except for ARM64).

If immintrin.h is included, it is guarded by `__SSE2__`, `__SSE3__`, `__AVX__`, `__AVX2__`, `ZEND_INTRIN_AVX2_NATIVE`, or `ZEND_INTRIN_AVX2_RESOLVER`. Not sure how to generally solve this, but maybe this PR should not use `HAVE_IMMINTRIN_H` header at all. although it would be possible to define that symbol on Windows (probably unconditionally).

And there is another issue, namely in hash_sha_ni.c and php_hash_sha.h we have `#if (defined(__i386__) || defined(__x86_64__)) && defined(HAVE_IMMINTRIN_H)`. but these are never defined by MSVC. Not sure if we have some portable symbols for that; otherwise `__WIN32` and `__WIN64` could be used (although the latter is also defined for ARM64 which is probably an issue).

Anyway, I've just skipped these checks for a test build, verified that `SHA256_Transform_shani()` is actually called, and ext/hash/tests/sha256.phpt passed. \o/

Reply:
> Ugh, that whole AVX detection is apparently pretty much messed up in php-src.

Yeah, that's what I meant by the open task item in the original issue: I have no idea what I was doing here and the existing source files where the intrinsics are used didn't help either, because they are preprocessor hell.

So I just did what looked reasonable and worked for me, hoping for someone knowledgeable to advice.